cmd/sts-handlers.go

```diff
@@ -593,3 +602,81 @@ func (sts *stsAPIHandlers) AssumeRoleWithLDAPIdentity(w http.ResponseWriter, r *
 
 	writeSuccessResponseXML(w, encodedSuccessResponse)
 }
+
+// AssumeRoleWithCertificate implements user authentication with client certificates.
+// It verifies the client-provided X.509 certificate, maps the certificate to an S3 policy
+// and returns temp. S3 credentials to the client.
+//
+// API endpoint: https://minio:9000?Action=AssumeRoleWithCertificate&Version=2011-06-15
+func (sts *stsAPIHandlers) AssumeRoleWithCertificate(w http.ResponseWriter, r *http.Request) {
+	var ctx = newContext(r, w, "AssumeRoleWithCertificate")
+
+	if !globalSTSTLSConfig.Enabled {
+		writeSTSErrorResponse(ctx, w, true, ErrSTSNotInitialized, errors.New("STS API 'AssumeRoleWithCertificate' is disabled"))
+		return
+	}
+
+	// We have to establish a TLS connection and the
+	// client must provide exactly one client certificate.
+	// Otherwise, we don't have a certificate to verify or
+	// the policy lookup would ambigious.
+	if r.TLS == nil {
+		writeSTSErrorResponse(ctx, w, true, ErrSTSInsecureConnection, errors.New("No TLS connection attempt"))
+		return
+	}
+	if len(r.TLS.PeerCertificates) == 0 {
+		writeSTSErrorResponse(ctx, w, true, ErrSTSMissingParameter, errors.New("No client certificate provided"))
+		return
+	}
+	if len(r.TLS.PeerCertificates) > 1 {
+		writeSTSErrorResponse(ctx, w, true, ErrSTSInvalidParameterValue, errors.New("More than one client certificate provided"))
+		return
+	}
+
+	var certificate = r.TLS.PeerCertificates[0]
+	if !globalSTSTLSConfig.InsecureSkipVerify {
+		_, err := certificate.Verify(x509.VerifyOptions{
+			KeyUsages: []x509.ExtKeyUsage{
+				x509.ExtKeyUsageClientAuth,
+			},
+			Roots: globalRootCAs,
+		})
+		if err != nil {
+			writeSTSErrorResponse(ctx, w, true, ErrSTSInvalidClientCertificate, err)
+			return
+		}
+	}
+
+	// We set the expiry of the temp. credentials to the minumim of the
+	// configured expiry and the duration until the certificate itself
+	// expires.
+	// We must not issue credentials that out-live the certificate.
+	var expiry = globalSTSTLSConfig.STSExpiry
+	if validUntil := time.Until(certificate.NotAfter); validUntil < expiry {
+		expiry = validUntil
+	}
+
+	tmpCredentials, err := auth.GetNewCredentialsWithMetadata(map[string]interface{}{
+		expClaim: time.Now().UTC().Add(expiry).Unix(),
+	}, globalActiveCred.SecretKey)
+	if err != nil {
+		writeSTSErrorResponse(ctx, w, true, ErrSTSInternalError, err)
+		return
+	}
+
+	// We map the X.509 subject common name to the policy. So, a client
+	// with the common name "foo" will be associated with the policy "foo".
+	// Other mapping functions - e.g. public-key hash based mapping - are
+	// possible but not implement.
+	//
+	// Group mapping is not possible with standard X.509 certificates.
+	if err = globalIAMSys.SetTempUser(tmpCredentials.AccessKey, tmpCredentials, certificate.Subject.CommonName); err != nil {
```

Comment:
should we validate that `CommonName` is empty or not?

Reply:
Doesn't `SetTempUser` check this? What happens if it is emtpy. Does the access key not a have any policy.

If not checked - we should.

Reply:
No, we need to enforce this at API level so that we expect that CommonName is always present. Expecting SetTempUser returning error doesn't have the context from the caller to indicate proper errors for the client.

Reply:
Also what is the TLS field that is non-optional? that can be used - IMO we should use that.

Reply:
Under this implementation, the ParentUser = Policy name associated with the user. This is a bit inflexible - you cannot assign multiple clients the same policy as they all need to have the same "CommonName" - is my understanding right?

Reply:
@harshavardhana The `Request.TLS` contains TLS status information (the TLS connection state) about the TLS connection over which the request has been sent/received. It contains, values like the SNI (server name that client has asked for), used ciphersuite, OCSP status information etc.
The entire `Request.TLS` field cannot be used. Imagine it like some status information about a TCP connection. It contains way to many volatile / unrecoverable data.

@donatello 
This is correct. However, there is no "standard" way to assign multiple clients (with different X.509 certificates) the same "value" derived from the certificate. The CN of each certificate must be unique. A CA must not issue the same CN for different clients.
The fact that there is a 1:1 mapping between client and policy is a known limitation.